container/src/state.rs

```diff
@@ -0,0 +1,222 @@
+use crate::{Error, Result};
+use serde::{Deserialize, Serialize};
+use std::fs;
+use std::fs::{File, OpenOptions};
+use std::io::Write;
+use std::path::PathBuf;
+use std::sync::{Arc, RwLock};
+
+const KAPS_ROOT_PATH: &str = "/var/run/kaps/containers";
+const OCI_VERSION: &str = "0.2.0";
+const STATE_FILE: &str = "state.json";
+
+/// Container runtime status
+#[derive(Serialize, Deserialize, Debug, PartialEq, Copy, Clone)]
+pub enum Status {
+    #[serde(rename = "creating")]
+    Creating,
+    #[serde(rename = "created")]
+    Created,
+    #[serde(rename = "running")]
+    Running,
+    #[serde(rename = "stopped")]
+    Stopped,
+}
+
+impl Default for Status {
+    fn default() -> Self {
+        Status::Creating
+    }
+}
+
+/// Represent the state of the running container.
+#[derive(Serialize, Deserialize, Debug)]
+pub struct ContainerState {
+    id: String,
+    /// OCI version.
+    oci_version: String,
+    /// Runtime state of the container.
+    pub status: Arc<RwLock<Status>>,
+    /// ID of the container process.
+    pub pid: i32,
+    /// Path to the bundle.
+    bundle: PathBuf,
+}
+
+impl Default for ContainerState {
+    fn default() -> Self {
+        ContainerState {
+            oci_version: OCI_VERSION.to_string(),
+            id: String::default(),
+            status: Arc::new(RwLock::new(Status::default())),
+            pid: 0,
+            bundle: PathBuf::default(),
+        }
+    }
+}
+
+impl ContainerState {
+    pub fn new(id: &str, bundle_path: &str) -> Result<Self> {
+        ContainerState::_new(id, bundle_path, KAPS_ROOT_PATH)
+    }
+
+    fn _new(id: &str, bundle_path: &str, container_dir: &str) -> Result<Self> {
+        let bundle = PathBuf::from(bundle_path);
+        let container_path = PathBuf::from(container_dir).join(id);
+
+        if container_path.as_path().exists() {
+            return Err(Error::ContainerExists(format!(
+                "A container with the id '{}' already exists",
+                id
+            )));
+        }
+
+        // create the container directory
+        fs::create_dir_all(&container_path).map_err(Error::CreateStateFile)?;
+
+        // create the `state.json` file of the container
+        File::create(container_path.join(STATE_FILE)).map_err(Error::CreateStateFile)?;
+
+        let container_state = ContainerState {
+            id: id.to_string(),
+            bundle,
+            ..Default::default()
+        };
+
+        container_state._save(container_dir)?;
+
+        Ok(container_state)
+    }
+
+    /// Get the current runtime status of the container.
+    ///
+    /// As The container status is a RwLock,
+    /// calling this function results in acquiring a read lock on the status.
+    fn _status(&self) -> Result<Status> {
+        let container_status = Arc::clone(&self.status);
+        let container_status = container_status
+            .read()
+            .map_err(|e| Error::StatusLockPoisoned(e.to_string()))?;
```

Comment:
As you said Error should take native type but this one uses lifetime and I'm not yet comfortable with this concept.

The difficulty comes in the déclaration of the Error in the `lib.rs` file, I'm not sure which lifetime ('a) to pass to the native error: `PoisonError<RwLockWriteGuard<'a, Status>>`.

Reply:
Ah, you'd need to add a lifetime to your Error, which would be an overkill. Let's keep it that way.